Add vectorized implementation

marcotama opened the pull request:
This is the vectorized implementation discussed in issue #4 .

marcotama committed:
Add vectorized implementation

encukou commented on bresenham_numpy.py:
```diff
@@ -0,0 +1,39 @@
+        x0: Union[float, int],
+        y0: Union[float, int],
+        x1: Union[float, int],
+        y1: Union[float, int]
```
The algorithm deals with integers, please don't allow `float`. (Or, if this is generalized to `float`, please test that case.)

marcotama replied:
I added tests for floats, but it is tricky.
Can you verify if the tests are satisfactory to you?
What I did is, I took some of the existing tests and moved the starting point by one unit in either X or Y (whichever has shorter delta between start and end point) towards the end point. Then I removed the first point from the results.

encukou commented on bresenham_numpy.py:
```diff
@@ -0,0 +1,39 @@
+    :param y0: The Y coordinate of the first point
+    :param x1: The X coordinate of the second point
+    :param y1: The Y coordinate of the second point
+    :return: The X and Y integer coordinates on the line from (x0, y0) to (x1, y1)
```
This should mention the return type (tuple of Numpy arrays).

marcotama replied:
Done!

encukou commented on requirements.txt:
```diff
@@ -0,0 +1 @@
+numpy==1.15.4
```
I'd be happier if this wasn't a hard requirement: the generator version is still useful without numpy, and should be usable (and tested) without numpy.

marcotama replied:
I removed the requirements.txt file

encukou commented:
Looks good! I'm adding comments in case you'd like to address/discuss them, but with a few changes I'll be happy to merge this implementation. Let me know if I should go ahead and do that.

marcotama committed:
Handle vertical lines and fix a bunch of problems

marcotama commented:
Also, I found a problem in the tests. The vectorized implementation basically just computes a line using floating point arithmetic and then calculates the pixel coordinates by rounding.
There is ambiguity when the line lies exactly on the boundary of two pixels. In those cases, I imagine the result depends on floating point errors.

In one of the existing tests (the one in `test_min_slope_two_way`), the original function gives a different solution compared to the vectorized function (specifically `(5, 0)` vs `(5, 1)`). The question is, can we document this and leave it? I am not sure how to go about making the two functions agree 100% in these circumstances.

encukou commented:
Well, the Bresenham algorithm (famously) uses only integers, so it should be exact. Can the numpy implementation use int as well?

marcotama commented:
I thought about it more and realized that floating point error does not play a role here. It's just that every algorithm must have some bias to solve ambiguous cases.

If you draw a line between `(0,0)` and `(2,1)`, you could color either `(1,0)` or `(1,1)`, and neither choice would be wrong because the "error" would be 0.5 in both cases.

Bresenham algorithm has a bias that depends on the direction of the line. Check out the tests that you wrote:
* When Bresenham algorithm draws a line from `(0,0)` to `(10,1)`, in the middle it chooses to color `(5,1)`
* When Bresenham algorithm draws a line from `(10,1)` to `(0,0)`, in the middle it chooses to color `(5,0)`
* When my algorithm draws a line from `(0,0)` to `(10,1)`, in the middle it chooses to color `(5,0)`
* When my algorithm draws a line from `(10,1)` to `(0,0)`, in the middle it chooses to color `(5,0)`

The behavior of my algorithm is due to the fact that the rounding operation is applied to the same number, `0.5`, in both cases. Numpy rounding has its own quirks: https://stackoverflow.com/questions/28617841/rounding-to-nearest-int-with-numpy-rint-not-consistent-for-5#28617911

In the end, these are all arbitrary choices/biases and, from a numerical perspective, all algorithms are equally correct. Maybe there are "looks-related" arguments for choosing the bias in Bresenham algorithm. Do you know if that is the case? I found a paper by Bresenham himself that elaborates on this (https://ieeexplore.ieee.org/abstract/document/4057178), but I do not have the time to dive deep into it; however, in the conclusions he says "Implementation-dependent resolutions are perhaps to be expected; we can hope then that any given implementation will document its particular choices...", so maybe we can resolve to document this behavior?

encukou commented:
I think we're talking past each other here.
The original bresenham algorithm only accepts integers, and works with integers. Floats and rounding should not come into play at all, and they should not be allowed as input.

If there's a well-researched generalization of the algorithm that allows floats, let's talk about adding it. But if it's about rounding the input up or down, why bother? The caller can do that quite easily.

encukou commented on bresenham_numpy.py:
```diff
@@ -0,0 +1,51 @@
+    else:
+        swapped = False
+
+    m = (y1 - y0) / (x1 - x0) if x1 - x0 != 0 else +1
```
I missed the float operation here. Sorry for that :(

marcotama commented:
No worries, as long as at some point we understand each other ;)

My reasoning is the following. Never mind whether rounding or floats are used in the original algorithm at all: the higher-level problem is coloring pixels so as to represent a line A-to-B on a integer space. Sometimes, the problem has more than one solution, such as in the case I mentioned before. The Bresenham algorithm has a certain intrinsic bias, tied to how x and y have asymmetric roles in the code. "my" algorithm has a different bias, tied to float rounding operations.

Even if we restrict the input type to integer, the ambiguity remains.

As for the generalization to float, going back to the mental model of drawing a line from A to B, you can imagine integers to point to the center of pixels, and floats to point to other points within a pixel.

Not sure I am addressing your issues yet :D

encukou commented:
> My reasoning is the following. Never mind whether rounding or floats are used in the original algorithm at all: the higher-level problem is coloring pixels so as to represent a line A-to-B on a integer space. 

But, if we abandon integers, it no longer makes much sense to use the name "bresenham". What I overlooked on the initial review was that in the end, it departs from the original idea quite a lot (see the leading paragraph of the [wiki page](https://en.wikipedia.org/wiki/Bresenham%27s_line_algorithm)).
Of course, if you're armed with floats and Numpy, the properties of Bresenham's algorithm aren't relevant to the task of drawing a line.

A new library called `rasterize` or `pixelize` or `gimme_coords`, with a hard dependency on Numpy, would make much more sense, don't you think? 

I'd be happy to link to it, or even set up the project and help with administrative overhead.

marcotama commented:
You make a good point! I agree with you, "my" algorithm is not Bresenham's, even if it solves the same problem.

Incidentally, in the meantime, I came across:
* a function part of `scikit-image`, [`skimage.draw.line`](http://scikit-image.org/docs/dev/api/skimage.draw.html#skimage.draw.line) that solves the same problem (twice as fast as my implementation!) and returns the same numbers as Bresenham's algorithm in the example above.
* a Python recipe (which I did not test) which also solves the rasterization problem but generalizes to N dimensions: http://code.activestate.com/recipes/578112-bresenhams-line-algorithm-in-n-dimensions/

I think, in light of this, I should leave it at that and avoid reinventing the wheel ;)

Maybe you could reference these two implementations in the README of the package?

encukou commented:
> Maybe you could reference these two implementations in the README of the package?

I [did just that](https://pypi.org/project/bresenham/0.2.1/). Sorry for the extreme delay – I had some matters of higher priority to attend to :)

marcotama commented:
No worries! Looks good! Just a small issue: on PyPi, in the line where you mention the Numpy recipe, the <var> tags are not handled correctly.

This is what I see:
<img width="616" alt="screen shot 2019-02-18 at 10 46 03 am" src="https://user-images.githubusercontent.com/1052500/52921383-83112680-336a-11e9-871a-851d27861658.png">

stefanv commented:
@marcotama We just merged an N-d line drawing algorithm into scikit-image, but I believe Bresenham would be better.  Are you interested in helping me out on the issue?  See https://github.com/scikit-image/scikit-image/pull/2043#issuecomment-535608255

marcotama commented:
Hi @stefanv , sorry about the long delay in response, life has been hectic! Am I too late to help?

stefanv commented:
@marcotama A version has been included in skimage, but I think it can be improved! Want to take a look at line_nd?

marcotama commented:
I put together a vectorized implementation of ND-Bresenham here: https://github.com/marcotama/scikit-image/commit/96c83ad49d4479476416449cbbd3ae59caf860be.

However there are a couple of problems:

1) in cases where either pixel is correct, Bresenham chooses the lower pixel while the original implementation chooses the higher (I discussed this kind of bias above: https://github.com/encukou/bresenham/pull/5#issuecomment-438872706). Is it ok to "break compatibility" this way?

2) If the algorithm is to work exclusively on integers, I am not entirely sure how to handle floats (currently truncating them). One idea I have is to use the decimal value to initialize the errors, but to do that I think I need to find the least common denominator of said decimals after converting them to fractions, which has two problems:
    a) it only works with fractional values (ok, on a computer no really real numbers exist, but then the LCD could have a huge denominator resulting in huge initial errors (think plotting from `(0,0)` to `(cos(30°), sin(30°))`, and
    b) computing the LCD sounds like extra work one might not want to do (but I guess we have the `integer` flag for that?).

What are your thoughts?

Also, should I open a PR while still experimenting or should I wait until stable? Or should I open an issue?

stefanv commented:
Thank you, @marcotama! Could you highlight why Bresenham is an improvement? (I guess it is a more optimal sampling of the line.) Bias, as long as the total does not increase, is not a problem, so feel free to break that: it was an arbitrary choice in the previous implementation.

I am not certain about your second question: are you referring to floating point start and end coordinates?

A PR with title "WIP:..." is typical for work in progress.

Thanks again for your contribution!

marcotama commented:
I opened a PR here: https://github.com/scikit-image/scikit-image/pull/4277
Not sure if it has any important advantage other than being floating point errors free.
